src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs

```diff
@@ -266,55 +265,62 @@ public void ScopeDispose_PreventsServiceResolution()
             //Check that resolution from root works
             Assert.NotNull(provider.CreateScope());
         }
-        
-        [ThreadStatic]
-        public static int ThreadId;
 
         [Fact]
-        public async Task ServiceProviderDispose_ContinueResolvingServices_Throws()
+        private void GetService_ThenDisposeOnDifferentThread_ServiceDisposed()
         {
             var services = new ServiceCollection();
-            ServiceProvider sp = null;
+            services.AddSingleton<Foo>();
+            var sp = services.BuildServiceProvider();
+            var foo = sp.GetRequiredService<Foo>();
+            Task.Run(() => sp.Dispose()).Wait();
+            
+            Assert.Equal(
+                Foo.TimesConstructorCalled,
+                Foo.TimesDisposeCalled);
+        }
 
-            var lazy = new Lazy<Thing1>(() =>
+        [Fact]
+        public void GetService_DisposeOnSameThread_Throws()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton<DisposableFoo>();
+            var sp = services.BuildServiceProvider();
+            Assert.Throws<ObjectDisposedException>(() =>
             {
-                // Tries to take the singleton lock (global)
-                return new Thing1(sp);
+                // foo ctor disposes ServiceProvider
+                var foo = sp.GetRequiredService<DisposableFoo>();
             });
+        }
 
-            services.AddTransient(sp =>
+        public class Foo : IDisposable
+        {
+            public static int TimesConstructorCalled = 0;
```

Comment:
Why is this static?

Reply:
Yeah it doesn't need to be static. I'm pushing up a cleanup for this diff.